fe/fe-core/src/main/java/com/starrocks/lake/ShardDeleter.java

```diff
@@ -87,35 +98,83 @@ private void deleteUnusedShard() {
 
             // 2. delete shard
             try {
-                GlobalStateMgr.getCurrentState().getStarOSAgent().deleteShards(shards);
+                starOSAgent.deleteShards(shards);
             } catch (DdlException e) {
                 LOG.warn("failed to delete shard from starMgr");
                 continue;
             }
+        }
+    }
 
-            deletedShards.addAll(shards);
+    // delete redundant shard/shardGroup
+    // collect all shardGroup from starMgr and do diff, the redundant groups expect 0 should be deleted from starMgr
+    // for shard groups which exist in starMgr but not fe, if their create time , delete tShardGroupInfo::getGroupIdhem
+    private void deleteUnusedShards() {
+        StarOSAgent starOSAgent = GlobalStateMgr.getCurrentState().getStarOSAgent();
+
+        // 1.1.get all partitions of fe
+        // 1.2.get all shard group from starMgr
+        // 1.3.compute diff
+        List<Long> groupIdFe = getAllPartitionId();
+        List<ShardGroupInfo> shardGroupInfos = starOSAgent.listShardGroup();
+        shardGroupInfos = shardGroupInfos.stream()
+                .filter(item -> item.getGroupId() != 0L)
+                .collect(Collectors.toList());
+
+        if (shardGroupInfos.isEmpty()) {
+            return;
         }
 
-        // 3. succ both, remove from the map
-        try (LockCloseable ignored = new LockCloseable(rwLock.writeLock())) {
-            shardIds.removeAll(deletedShards);
+        // for debug
+        LOG.info("size of groupIdFe is {}, size of shardGroupInfos is {}",
+                groupIdFe.size(), shardGroupInfos.size());
+        LOG.info("groupIdFe is {}", groupIdFe);
+
+        Map<Long, String> groupToCreateTimeMap =
+                shardGroupInfos.stream().collect(Collectors.toMap(
+                        obj -> obj.getGroupId(),
+                        obj -> obj.getPropertiesMap().get("createTime"),
+                        (key1, key2) -> key1
+                ));
+
+        List<Long> groupIdStaros = shardGroupInfos.stream().map(ShardGroupInfo::getGroupId).collect(Collectors.toList());
+        List<Long> diff = groupIdStaros.stream().filter(e -> {
+            return !groupIdFe.contains(e);
+        }).collect(Collectors.toList());
+
+        // for debug
+        LOG.info("diff.size is {}", diff.size());
+        LOG.info("diff is {}", diff);
+
+        // 1.4.collect redundant shard groups and delete
+        List<Long> emptyShardGroup = new ArrayList<>();
+        for (long groupId : diff) {
+            if (System.currentTimeMillis() - Long.valueOf(groupToCreateTimeMap.get(groupId))
+                    > Config.shard_group_clean_interval) {
+                try {
+                    List<Long> shardIds = starOSAgent.listShard(groupId);
+                    if (shardIds.isEmpty()) {
+                        emptyShardGroup.add(groupId);
+                    } else {
+                        dropTabletAndDeleteShard(shardIds, starOSAgent);
```

Comment:
what if the tablet deletion is failed and shard deletion is successful? looks like the tablet never has a chance to be deleted again in future rounds.